testsupport/wait/host.go

```diff
@@ -162,138 +168,158 @@ func (a *HostAwaitility) UpdateUserSignupSpec(userSignupName string, modifyUserS
-			mur.Name, mur.Status.ProvisionedTime.String(), expectedTime.String())
-		return false
+	return MasterUserRecordWaitCriterion{
+		Match: func(a *HostAwaitility, actual *toolchainv1alpha1.MasterUserRecord) bool {
+			return actual.Status.ProvisionedTime != nil && expectedTime.Time == actual.Status.ProvisionedTime.Time
+		},
+		Diff: func(actual *toolchainv1alpha1.MasterUserRecord) string {
+			if actual.Status.ProvisionedTime == nil {
+				return fmt.Sprintf("expected status provisioned time '%s'.\n\tactual is 'nil'", expectedTime.String())
+			}
+			return fmt.Sprintf("expected status provisioned time '%s'.\n\tactual: '%s'", expectedTime.String(), actual.Status.ProvisionedTime.Time.String())
+		},
 	}
 }
 
 // UntilMasterUserRecordHasCondition checks if MasterUserRecord status has the given conditions (among others)
-func UntilMasterUserRecordHasCondition(condition toolchainv1alpha1.Condition) MasterUserRecordWaitCriterion {
-	return func(a *HostAwaitility, mur *toolchainv1alpha1.MasterUserRecord) bool {
-		if test.ContainsCondition(mur.Status.Conditions, condition) {
-			a.T.Logf("status conditions match in MasterUserRecord '%s`", mur.Name)
-			return true
-		}
-		a.T.Logf("waiting for status condition of MasterUserRecord '%s'. Actual: '%+v'; Expected: '%+v'", mur.Name, mur.Status.Conditions, condition)
-		return false
+func UntilMasterUserRecordHasCondition(expected toolchainv1alpha1.Condition) MasterUserRecordWaitCriterion {
+	return MasterUserRecordWaitCriterion{
+		Match: func(a *HostAwaitility, actual *toolchainv1alpha1.MasterUserRecord) bool {
+			return test.ContainsCondition(actual.Status.Conditions, expected)
+		},
+		Diff: func(actual *toolchainv1alpha1.MasterUserRecord) string {
+			return fmt.Sprintf("expected conditions to contain: %s.\n\tactual: %s", spew.Sdump(expected), spew.Sdump(actual.Status.Conditions))
+		},
 	}
 }
 
 // UntilMasterUserRecordHasConditions checks if MasterUserRecord status has the given set of conditions
-func UntilMasterUserRecordHasConditions(conditions ...toolchainv1alpha1.Condition) MasterUserRecordWaitCriterion {
-	return func(a *HostAwaitility, mur *toolchainv1alpha1.MasterUserRecord) bool {
-		if test.ConditionsMatch(mur.Status.Conditions, conditions...) {
-			a.T.Logf("status conditions match in MasterUserRecord '%s`", mur.Name)
-			return true
-		}
-		a.T.Logf("waiting for status condition of MasterUserRecord '%s'. Actual: '%+v'; Expected: '%+v'", mur.Name, mur.Status.Conditions, conditions)
-		return false
+func UntilMasterUserRecordHasConditions(expected ...toolchainv1alpha1.Condition) MasterUserRecordWaitCriterion {
+	return MasterUserRecordWaitCriterion{
+		Match: func(a *HostAwaitility, actual *toolchainv1alpha1.MasterUserRecord) bool {
+			return test.ConditionsMatch(actual.Status.Conditions, expected...)
+		},
+		Diff: func(actual *toolchainv1alpha1.MasterUserRecord) string {
+			return fmt.Sprintf("expected conditions to match:\n%s", Diff(actual.Status.Conditions, expected))
+		},
```

Comment:
why do they use different ways of printing the diff when they both verify conditions?

Reply:
the `UntilMasterUserRecordHasCondition` expects the resources conditions to contain the given `expected` condition (among others), whereas `UntilMasterUserRecordHasConditions` expects a full match. But I'm not sure if this answers your question

Reply:
TBH, it's very confusing when it should use/print the diff and when the full dump.

Reply:
it's implementation details ;) The idea is that the test will call the `Diff` func and to display the message when one or more condiitions did not match, and provide the most accurate message to help understand what did not match

Reply:
I see. I know that it's an implementation detail but we will be the ones who will implement it - it's not really clear to me in which cases it should print the dump and in which the diff.

Reply:
I mean - wouldn't it be easier to print the actual content vs the expected one? That logic could be fully generic and it could provide functions like:
```go
PrintExpectedVsActual(expected, actual interface{})
PrintActualThatDoesNotContain(actual, shouldContain interface{})
```
all the objects can be marshaled either as a YAML or as a  JSON, right? If we use correct indentation, then such contents are very easy to compare.